### DISCUSSION TITLE
Authorization example using directive

### DISCUSSION THREAD

**Discussion by theprobugmaker** [99designs/gqlgen#1875] [Q&A] [answered]
Is there an example of how to use gqlgen with authorization in a directive? I'm receiving a JWT token in the request. I'm using `gin-gonic` as well.

I want to check if the user permissions match the ones in the directive.

Thank you very much

**Comment by txbrown** [comment, marked as answer]
@theprobugmaker I believe I created mine based on the official docs [here](https://gqlgen.com/reference/directives/) 

If you follow the example you should be able to arrive to this.

```go
	c := generated.Config{
		Resolvers: gql.NewResolver(),
	}

	// Schema Directive - @isAuthenticated
	c.Directives.IsAuthenticated = func(ctx context.Context, obj interface{}, next graphql.Resolver) (res interface{}, err error) {
		ctxUserID := ctx.Value(utils.ProjectContextKeys.UserCtxKey)
		if ctxUserID == nil {
			return nil, common.GqlUnauthorizedError(ctx)
		}
		return next(ctx)
	}

	h := handler.GraphQL(generated.NewExecutableSchema(c))

	return func(c *gin.Context) {
		h.ServeHTTP(c.Writer, c.Request)
	}
	```
	
	on your middleware that checks token you should also inject the user in the context so this directive and your resolvers have access to it. I assume you have this done too, otherwise happy to share more code.

**Reply by wilhelmeek** [reply]
This is bang on. @theprobugmaker if your directive has arguments they will appear in the type signature (after generating) which you can use to match properties on your request context.

**Reply by lcucurachi** [reply]
I've noticed these directives are executed after the handler processes the data in the resolver. This is basically a bad approach... Authorisation should be handled in middleware and before resolvers prepare/process the data. Is there a way to configure gqlgen to do that? The generated.go file is as the name suggests "generated" so any modification is pointless unless you want to maintain it completely yourself which goes agains the point of the library.

**Comment by amen160** [comment]
We only want get JWT token from header. How to do it? @theprobugmaker

**Reply by txbrown** [reply]
You can find pretty good working examples here and adjust to your usecase
https://github.com/cmelgarejo/go-gql-server/blob/master/internal/handlers/auth/middleware/auth.go
